src/src_main/crypto_tc.c

```diff
@@ -428,31 +467,56 @@ int32_t Crypto_TC_ApplySecurity(const uint8_t* p_in_frame, const uint16_t in_fra
         */
         // TODO: Revisit this
         // TODO: Likely SA API Call
-        for (i = 0; i < sa_ptr->shplf_len; i++)
+        /* 4.1.1.5.2 The Pad Length field shall contain the count of fill bytes used in the
+        ** cryptographic process, consisting of an integral number of octets. - CCSDS 3550b1
+        */
+        // TODO: Set this depending on crypto cipher used
+
+        if(padding_needed)
         {
-            /* 4.1.1.5.2 The Pad Length field shall contain the count of fill bytes used in the
-            ** cryptographic process, consisting of an integral number of octets. - CCSDS 3550b1
-            */
-            // TODO: Set this depending on crypto cipher used
-            *(p_new_enc_frame + index) = 0x00;
-            index++;
+            char hex_padding[3] = {0};  //TODO: Create #Define for the 3
+            padding_needed = padding_needed & 0x00FFFFFF; // Truncate to be maxiumum of 3 bytes in size
+            
+            // Byte Magic
+            hex_padding[0] = (padding_needed >> 16) & 0xFF;
+            hex_padding[1] = (padding_needed >> 8)  & 0xFF;
+            hex_padding[2] = (padding_needed)  & 0xFF;
+            
+
+            uint8_t padding_start = 0;
+            padding_start = 3 - sa_ptr->shplf_len;
+
+            for (i = 0; i < sa_ptr->shplf_len; i++)
```

Comment:
This for loop condition doesn't seem correct -- the shplf_len field is the length of the pad_length field in the header, whereas this loop is filling in the actual padding in the frame body... They are not related.

You probably want the condition to be related to the hex-padding char array?

Reply:
I actually think this one is ok.
Yes, it's using the pad_length field, which according to the standard can only ever be a max of 3 bytes.
In this case, it is one byte (as that is all that will ever be needed to say 16 bytes of padding are necessary).
So, what is happening here, is that it is writing the padding into the pad length field in the header.

So, if 2 bytes of padding were necessary, it would write a 2 into the frame.  
Later, it will write the two bytes of 0202 into the actual data.

Reply:
You are confusing sa_ptr->shplf_len and p_new_enc_frame[INDEX_OF_PAD_LENGTH_FIELD]

Reply:
sa_ptr->shplf_len will almost always be 1.

Reply:
Is this loop (line 489) just adding the required padding bytes as necessary to the SDLS header which is variable (0-3)?

Reply:
It could be up to 3 though, correct?

Reply:
That's where the padding_start variable comes into play.  This whole piece of code is just breaking however much padding is needed into individual hex values, and then writing it into p_new_enc_frame[INDEX_OF_PAD_LENGTH_FIELD].

the sa_ptr->shplf_len value in the for loop is just making sure that we only write the number of bytes specified in the security association.

Reply:
Think about it like that, what does sa_ptr->shivf_len represent? It's the length of the IV in the security header, usually 12 bytes --- It does not represent the actual value of the IV.

What does sa_ptr->shplf_len represent? It's the length of the pad-field in the security header, usually 1 byte -- It does not represent the actual value of the pad length.

Reply:
That does not represent the actual padding either, though.
It just represents how many bytes of padding there will be.

Reply:
Right, what I'm saying is sa_ptr->shplf_len will never represent how many bytes of padding there will be.

Reply:
And the representation of how many bytes of padding that are needed, could be a value up to 3 bytes in length.

Reply:
How do you remove the padding bytes, then?

Reply:
Or rather, where is that 1 byte of data written?

Reply:
Maybe this one merits a voice conversation.
However, just to recap.
The security association sets the sa_ptr->shplf_len.
For AES256_CBC, it is 1.  It is 1, because within 1 byte, we can represent up to the number 16.  16 bytes of padding that could be necessary - 16, because if you send a block of 16, it will pad that 16 with an additional 16 bytes of padding per the standard.

Which means that at  p_new_enc_frame[INDEX_OF_PAD_LENGTH_FIELD],
There will be 1 byte that represents however much padding the data to be encrypted needs.

However, shplf_len, technically could be any value between 0-3, inclusively.
So, that's what this code is doing.  It is handling the case that we add additional ciphers in the future that need different shplf_len values greater than 1.  Which does become unlikely if we only use 1024 max lengths.

Reply:
So after webexing -- I realize that I misunderstood what this code block was doing -- I thought you're adding the actual padding here, but actually you're just setting the length in the SDLS header. This can be disregarded...